### DISCUSSION TITLE
Secure remote password protocol support

### DISCUSSION THREAD

**Discussion by korneliuscode** [zitadel/zitadel#7479] [Ideas]
Hello, 
Is there any way to integrate SRP into zitadel? This would be another very strong addition.
More details on SRP here: https://en.m.wikipedia.org/wiki/Secure_Remote_Password_protocol

**Comment by hifabienne** [comment]
Hei @korneliuscode 
At the moment there is no way of integrating SRP into ZITADEL.
@livio-a and @muhlemmer might have some opinions about that topic.

**Comment by muhlemmer** [comment]
SRP is a form off challenge-response authentication. We already support webauthN technology, where a hardware token signs a challenge. Some tokens offer PIN protection, (like yubikey,windows Hello, using the system's TPM chip) which enables "passwordless" authentication.  IMO that technology is far more superior as password based challenge response. So why would you still need SRP?

Also, there is already an issue open for the new [OAuth 2.0 Demonstrating Proof-of-Possession](https://github.com/zitadel/zitadel/issues/5402) which is yet another challenge-response method that does not require passwords to be send over the wire.

The wikipedia article mentions some benefits of SRP, and I will counter-argument them here:

> Like all PAKE protocols, an eavesdropper or [man in the middle](https://en.m.wikipedia.org/wiki/Man-in-the-middle_attack) cannot obtain enough information to be able to [brute-force guess](https://en.m.wikipedia.org/wiki/Brute-force_attack) a password or apply a [dictionary attack](https://en.m.wikipedia.org/wiki/Dictionary_attack) without further interactions with the parties for each guess. Furthermore, being an augmented PAKE protocol, the server does not store password-equivalent data.[[2]](https://en.m.wikipedia.org/wiki/Secure_Remote_Password_protocol#cite_note-2) This means that an attacker who steals the server data cannot masquerade as the client unless they first perform a brute force search for the password.

We use the latest TLS protocols to protect us against MITM attacks. If a flaw in TLS would allow such an attack, there is a bigger problem, as we use the same channels to return tokens to applications. So the password might be safe, but the user and the application is compromised regardless. Hence, we have to assume TLS is safe and resistant against MITM. If such exposure happens, the attacker is also free to set its own credentials for the user, using the stolen tokens.

> Furthermore, the server also needs to know about the password (but not the password itself) in order to instigate the secure connection.

Again, we already use TLS there is no additional benefit here.

> it allows a user to authenticate themselves to a server, it is resistant to [dictionary attacks](https://en.m.wikipedia.org/wiki/Dictionary_attack) mounted by an eavesdropper,

Zitadel uses strong hashing of passwords with globally unique random salts. If there ever would be a data leak, the stored hashes are more or less useless to the attacker.

> and it does not require a [trusted third party](https://en.m.wikipedia.org/wiki/Trusted_third_party).

We build trust by sharing our source code. You can checkout our [password hashing library](https://github.com/zitadel/passwap) and see for yourself that we adhere to best practices when it comes to hashing and storing passwords.

> It effectively conveys a [zero-knowledge password proof](https://en.m.wikipedia.org/wiki/Zero-knowledge_password_proof) from the user to the server

----

To conclude, I do not think that implementing SRP significantly reduces attack surface for it to be valuable to our users, given the fact that we already have a strong alternative (WebAuthN) and another one proposed (DPoP).

**Comment by korneliuscode** [comment]
Thanks for sharing. Looks like there are different opinions. We still see plenty customer/user scenarios where passwords are the „preferred way“, independent of other upcoming or technical possibilities. 
Therefore it would still be a great addition to the platform.
